### PR TITLE
More test on pkg/ioutils (coverage)

### LINKED ISSUE #11595: pkg/ioutils - Test coverage
**Problem**
 
pkg/ioutils - Test coverage
 
**Solution**
 
The `pkg/ioutils` package is used at multiple places in the Docker daemon, yet  the package test coverage could be improved (it's 51% at the time of filing this issue).
 
Improve test coverage on the `pkg/ioutils` package. You can use the following commands to measure progress:
 
    cd pkg/ioutils
    go test -cover -coverprofile /dev/stderr 2>&1 >/dev/null | go tool cover -func /dev/stdin

### PR DESCRIPTION
**Pull request by vdemeester**
Add tests on :
- `pkg/ioutils/readers.go`
- `pkg/ioutils/writers.go`

Closes #11595

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

File: pkg/ioutils/readers_test.go [hunk 1]
```diff
@@ -2,11 +2,94 @@ package ioutils
...
 
+// Implement io.Reader
+type errorReader struct {
+}
+
+func (r *errorReader) Read(p []byte) (int, error) {
+	return 0, fmt.Errorf("Error reader always fail.")
+}
+
+func newErrorReader() io.Reader {
```
Comment (tiborvass) [inline pkg/ioutils/readers_test.go]: no need for this constructor, might as well inline it

File: pkg/ioutils/readers_test.go [hunk 2]
```diff
@@ -2,11 +2,94 @@ package ioutils
...
 	"bytes"
+	"fmt"
 	"io"
 	"io/ioutil"
+	"strings"
 	"testing"
 )
 
+// Implement io.Reader
+type errorReader struct {
```
Comment (tiborvass) [inline pkg/ioutils/readers_test.go]: you can write this on one line: `type errorReader struct{}`

File: pkg/ioutils/readers_test.go [hunk 3]
```diff
@@ -2,11 +2,94 @@ package ioutils
...
+	return &errorReader{}
+}
+
+func TestReadCloserWrapperClose(t *testing.T) {
+	reader := strings.NewReader("A string reader")
+	wrapper := NewReadCloserWrapper(reader, func() error {
+		return fmt.Errorf("This will be called when closing")
+	})
+	err := wrapper.Close()
+	if err == nil {
```
Comment (tiborvass) [inline pkg/ioutils/readers_test.go]:
```
if err := ...; err == nil {
}
```

File: pkg/ioutils/writers_test.go [hunk 4]
```diff
@@ -6,6 +6,32 @@ import (
...
 
+func TestWriteCloserWrapperClose(t *testing.T) {
+	called := false
+	writer := bytes.NewBuffer([]byte{})
+	wrapper := NewWriteCloserWrapper(writer, func() error {
+		called = true
+		return nil
+	})
+	err := wrapper.Close()
+	if err != nil {
```
Comment (tiborvass) [inline pkg/ioutils/writers_test.go]: `if err := ...; err != nil {`

**Comment by tiborvass** [comment]
@vdemeester apart from some small style changes, it's great to increase coverage, thanks!

File: pkg/ioutils/writers_test.go (hunk 4 above)
Comment (vdemeester) [inline pkg/ioutils/writers_test.go]: I always forget I can do that in go :)

File: pkg/ioutils/readers_test.go (hunk 3 above)
Comment (vdemeester) [inline pkg/ioutils/readers_test.go]: All right, will do.

**Comment by vdemeester** [comment]
@tiborvass took care of your comments I think :)

File: pkg/ioutils/writers_test.go (hunk 4 above)
Comment (runcom) [inline pkg/ioutils/writers_test.go]: or `c.Assert(wrapper.Close(), check.IsNil)` (just learnt thanks to @tiborvass :smile: )

Comment (vdemeester) [inline pkg/ioutils/writers_test.go]: @runcom yeah but these test are not (yet) using gocheck, so can't do for now.. But i'll definitely look at how to use gocheck for these tests too (not for the PR :)).

Comment (runcom) [inline pkg/ioutils/writers_test.go]: @vdemeester forget what I said, I was jus happy to share that :smile:

Comment (vdemeester) [inline pkg/ioutils/writers_test.go]: @runcom and I am happy  you shared :wink:

**Comment by tiborvass** [comment]
@vdemeester there's this one still https://github.com/docker/docker/pull/12954/files#r29599616

**Comment by vdemeester** [comment]
@tiborvass Ah missed it >_<

**Comment by LK4D4** [comment]
Windows fail is unrelated.
LGTM

**Comment by tiborvass** [comment]
LGTM

**Comment by tiborvass** [comment]
Thanks for your contribution @vdemeester !

**Pull request merged**